buildSrc/src/main/java/com/google/firebase/gradle/plugins/UpdatePinnedDependenciesTask.kt

```diff
@@ -0,0 +1,115 @@
+import org.gradle.api.tasks.StopExecutionException
+import org.gradle.api.tasks.TaskAction
+
+/**
+ * Changes the project level dependencies in a project's build file to pinned dependencies.
+ *
+ * Computes project level dependencies outside of the project's library group, looks for instances
+ * of them in the `dependencies` block in the build file, and replaces them with pinned versions.
+ * GMaven is used to figure out the latest version to use for a given dependency.
+ *
+ * For example, given the following input:
+ * ```
+ * dependencies {
+ *    implementation(project(":appcheck:firebase-appcheck-interop"))
+ * }
+ * ```
+ *
+ * The output would be:
+ * ```
+ * dependencies {
+ *    implementation("com.google.firebase:firebase-appcheck-interop:17.0.1")
+ * }
+ * ```
+ *
+ * *Assuming that `17.0.1` is the latest version of `firebase-appcheck-interop`*
+ *
+ * @property buildFile A [File] that should be used as the source to update from. Typically the
+ * `build.gradle` or `build.gradle.kts` file for a given project.
+ * @property outputFile A [File] that should be used to write the new text to. Typically the same as
+ * the input file ([buildFile]).
+ * @see PostReleasePlugin
+ */
+abstract class UpdatePinnedDependenciesTask : DefaultTask() {
+  @get:[InputFile]
+  abstract val buildFile: Property<File>
+
+  @get:[OutputFile]
+  abstract val outputFile: Property<File>
+
+  @TaskAction
+  fun build() {
+    val needsChanging = computeLibrariesThatNeedChanging()
+
+    val lines = buildFile.get().readLines()
+    val newLines = updateDependencyLines(lines, needsChanging)
+
+    if (newLines == lines) throw missingProjectLevelDependenciesError(needsChanging)
+
+    outputFile.get().writeText(newLines.joinToString("\n") + "\n")
+  }
+
+  private fun computeLibrariesThatNeedChanging(): List<FirebaseLibraryExtension> {
+    val firebaseLibrary = project.firebaseLibrary
+    val sisterProjects = firebaseLibrary.getLibrariesToRelease()
+    val dependencies = projectLevelDependenciesForLibrary(firebaseLibrary)
+    val needsChanging = dependencies - sisterProjects
+
+    if (needsChanging.isEmpty()) throw StopExecutionException("No libraries to change.")
+
+    return needsChanging
+  }
+
+  private fun projectLevelDependenciesForLibrary(library: FirebaseLibraryExtension) =
+    library.resolveProjectLevelDependencies().filterNot { it.path in DEPENDENCIES_TO_IGNORE }
+
+  private fun updateDependencyLines(
+    lines: List<String>,
+    libraries: List<FirebaseLibraryExtension>
+  ) =
+    lines.replaceMatches(DEPENDENCY_REGEX) {
+      val projectName = it.firstCapturedValue
+      val projectToChange = libraries.find { it.path == projectName }
+      val latestVersion = projectToChange?.latestVersion
+
+      latestVersion?.let { "\"${projectToChange.mavenName}:$latestVersion\"" }
+    }
+
+  // This should never occur, but if it does- it's probably a regex error. Better safe than sorry?
+  private fun missingProjectLevelDependenciesError(
```

Comment:
I'd rather not wrap the exception throwing inside a function since it obscures where errors are found.

Reply:
SGTM. Mainly did it to abstract the long line and help with readability.